File: apps/go/chapter07/strings_test.go
package chapter07_test

import (
	"testing"

	"github.com/k2works/getting-started-algorithm/go/chapter07"
)

func TestBfMatch(t *testing.T) {
	got := chapter07.BfMatch("ABCXDEZCABACABAB", "ABAB")
	if got != 12 {
		t.Errorf("BfMatch(...) = %d; want 12", got)
	}
	got = chapter07.BfMatch("ABCXDEZCABACABAB", "XYZ")
	if got != -1 {
		t.Errorf("BfMatch(no match) = %d; want -1", got)
	}
}

func TestKmpMatch(t *testing.T) {
	got := chapter07.KmpMatch("ABCXDEZCABACABAB", "ABAB")
	if got != 12 {
		t.Errorf("KmpMatch(...) = %d; want 12", got)
	}
}

func TestBmMatch(t *testing.T) {
	got := chapter07.BmMatch("ABCXDEZCABACABAB", "ABAB")
	if got != 12 {
		t.Errorf("BmMatch(...) = %d; want 12", got)
	}
}

func TestCountChars(t *testing.T) {
	got := chapter07.CountChars("hello")
	expected := map[rune]int{'h': 1, 'e': 1, 'l': 2, 'o': 1}
	for k, v := range expected {
		if got[k] != v {
			t.Errorf("CountChars('hello')[%c] = %d; want %d", k, got[k], v)
		}
	}
}

func TestReverseString(t *testing.T) {
	got := chapter07.ReverseString("hello")
	if got != "olleh" {
		t.Errorf("ReverseString('hello') = %q; want 'olleh'", got)
	}
}

func TestIsPalindrome(t *testing.T) {
	if !chapter07.IsPalindrome("racecar") {
		t.Error("IsPalindrome('racecar') should be true")
	}
	if chapter07.IsPalindrome("hello") {
		t.Error("IsPalindrome('hello') should be false")
	}
}

File: apps/c/chapter07/strings_test.c
#include <string.h>
#include "../include/test_helper.h"
#include "strings.h"

void test_bf_match(void) {
    ASSERT_EQ_INT(12, bf_match("ABCXDEZCABACABAB", "ABAB"));
    ASSERT_EQ_INT(-1, bf_match("ABCXDEZCABACABAB", "XYZ"));
}

void test_kmp_match(void) {
    ASSERT_EQ_INT(12, kmp_match("ABCXDEZCABACABAB", "ABAB"));
    ASSERT_EQ_INT(-1, kmp_match("ABCXDEZCABACABAB", "XYZ"));
}

void test_bm_match(void) {
    ASSERT_EQ_INT(12, bm_match("ABCXDEZCABACABAB", "ABAB"));
    ASSERT_EQ_INT(-1, bm_match("ABCXDEZCABACABAB", "XYZ"));
}

void test_count_chars(void) {
    CharCount cc[26];
    int n = count_chars("hello", cc, 26);
    /* find 'l' */
    int found = 0;
    for (int i = 0; i < n; i++) {
        if (cc[i].c == 'l') { ASSERT_EQ_INT(2, cc[i].count); found = 1; }
    }
    ASSERT_TRUE(found);
}

void test_reverse_string(void) {
    char buf[16];
    reverse_string("hello", buf, sizeof(buf));
    ASSERT_EQ_STR("olleh", buf);
}

void test_is_palindrome(void) {
    ASSERT_TRUE(is_palindrome("racecar"));
    ASSERT_FALSE(is_palindrome("hello"));
}

int main(void) {
    test_bf_match();
    test_kmp_match();
    test_bm_match();
    test_count_chars();
    test_reverse_string();
    test_is_palindrome();
    TEST_SUMMARY();
    return 0;
}
